scripted-diff: Regenerate key_io data deterministically

maflcko opened the pull request.

MarcoFalke commented:
Requested by @Sjors here: https://github.com/bitcoin/bitcoin/pull/20861#pullrequestreview-596345213

Commit:
scripted-diff: Regenerate key_io data deterministically

-BEGIN VERIFY SCRIPT-
 ./contrib/testgen/gen_key_io_test_vectors.py valid 70 > ./src/test/data/key_io_valid.json
 ./contrib/testgen/gen_key_io_test_vectors.py invalid 70 > ./src/test/data/key_io_invalid.json
-END VERIFY SCRIPT-

Commit:
contrib: make gen_key_io_test_vectors deterministic

Also, remove instructions which are redundant with the README

Sjors commented:
tACK fab2a4a, modulo the numbers in the `README` should be updated...

Also, I find it less tedious to run this from the root of the project:
```
PYTHONPATH=test/functional/test_framework contrib/testgen/gen_key_io_test_vectors.py valid 61 > src/test/data/key_io_valid.json
PYTHONPATH=test/functional/test_framework contrib/testgen/gen_key_io_test_vectors.py invalid 60 > src/test/data/key_io_invalid.json
git diff

```

MarcoFalke commented:
Done

Sjors commented on contrib/testgen/README.md:
```diff
@@ -2,7 +2,9 @@
 
-    PYTHONPATH=../../test/functional/test_framework ./gen_key_io_test_vectors.py valid 50 > ../../src/test/data/key_io_valid.json
-    PYTHONPATH=../../test/functional/test_framework ./gen_key_io_test_vectors.py invalid 50 > ../../src/test/data/key_io_invalid.json
+    -BEGIN VERIFY SCRIPT-
```
The `-BEGIN VERIFY SCRIPT-` bit is confusing.

MarcoFalke replied:
fixed

sipa commented:
This is generally a good idea, and we should do it. But this does still mean that changes to what is generated based on the RNG output will result in an entirely different output. E.g. adding an option to a random.choice() may cause more entropy to be extracted from the RNG, and thus cause all future data from it to be shifted. In theory this could be addresses using consistent hashing, but that's probably overkill.

Concept ACK

Sjors commented:
re-utACK fa21feb0b9c93fbc205d102510d3bb7a474431c5

MarcoFalke commented on contrib/testgen/gen_key_io_test_vectors.py:
```diff
@@ -4,14 +4,9 @@
-    PYTHONPATH=../../test/functional/test_framework ./gen_key_io_test_vectors.py valid 50 > ../../src/test/data/key_io_valid.json
-    PYTHONPATH=../../test/functional/test_framework ./gen_key_io_test_vectors.py invalid 50 > ../../src/test/data/key_io_invalid.json
 '''
 # 2012 Wladimir J. van der Laan
```
@laanwj This says your name, which makes you qualified to review this? :grimacing:

laanwj replied:
I definitely don't remember writing this :sweat:

MarcoFalke replied:
Your name is removed in current master, but you are still welcome to review

MarcoFalke commented:
Rebased

Sjors commented:
ACK fa506add25cbe5efbbabca647f5378c4128cf945

I get the same JSON files when running the script. There are no duplicate addresses and they all change if you modify the `42` seed value (as expected).

laanwj commented:
Tested ACK fa506add25cbe5efbbabca647f5378c4128cf945
Good idea to put it in the scripted-diff. I also verified locally that the output is the same.
